src/mlpack/methods/ann/layer/lambda_map_reduce.hpp

```diff
@@ -0,0 +1,270 @@
+/**
+ * @file methods/ann/layer/lambda_map_reduce.hpp
+ * @author Shubham Agrawal
+ *
+ * Base class for neural network layers that are wrappers around other layers.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_LAMBDA_MAP_REDUCE_HPP
+#define MLPACK_METHODS_ANN_LAYER_LAMBDA_MAP_REDUCE_HPP
+
+#include "../make_alias.hpp"
+#include "layer.hpp"
+#include <mlpack/methods/ann/reduction_rules/add_reduction.hpp>
+
+namespace mlpack {
+namespace ann {
+
+/**
+ * A lambda "map-reduce" is a layer that is a wrapper around other layers.  
+ * It passes the input through all of its child layers sequentially, returning
+ * the output from reducing the output.
+ *
+ * It's likely not very useful to use this layer directly; instead, this layer
+ * is meant as a base class for use by other layers that must store and use
+ * multiple layers.
+ *
+ * @tparam ReductionRuleType Reduction rule to use for the layer.
+ * @tparam MatType Matrix representation to accept as input and use for
+ *    computation.
+ */
+template<typename ReductionRuleType, typename MatType>
+class LambdaMapReduceType : public Layer<MatType>
```

Comment:
Hmm, probably you could inherit from `MultiLayer` here instead, and use the functions there that allocate memory for each layer's outputs, etc.

Reply:
No, I think I need to override the private memory allocation method. That's why I inherited Layer instead of MultiLayer.

Reply:
I thought so too for `Concat` but it turned out to not be fully necessary.  I'll post a link to the relevant bits of the PR that I open later today.